lib/GIL/BasicBlock.cpp

```diff
@@ -0,0 +1,131 @@
+#include "BasicBlock.hpp"
+#include <algorithm>
+#include <iterator>
+
+namespace glu::gil {
+
+BasicBlock::BasicBlock()
+{
+}
+
+BasicBlock::BasicBlock(
+    std::string const &label, std::list<InstBase *> const &instructions
+)
+    : _instructions(std::move(instructions))
+    , _label(std::move(label))
+{
+}
+
+BasicBlock::BasicBlock(std::list<InstBase *> const &instructions)
+    : _instructions(std::move(instructions))
+{
+}
+
+std::list<InstBase *> BasicBlock::getInstructions() const
+{
+    return _instructions;
+}
+
+std::size_t BasicBlock::getInstructionCount() const
+{
+    return _instructions.size();
+}
+
+InstBase *BasicBlock::popFirstInstruction()
+{
+    if (!_instructions.empty()) {
+        auto inst = _instructions.front();
+        _instructions.pop_front();
+        return inst;
+    }
+
+    return nullptr;
+}
+
+InstBase *BasicBlock::getInstructionAt(std::size_t index) const
+{
+    if (index < _instructions.size()) {
+        auto it = _instructions.begin();
+        std::advance(it, index);
+        return *it;
+    }
+
+    return nullptr;
+}
+
+void BasicBlock::addInstructionAtEnd(InstBase *inst)
+{
+    _instructions.push_back(inst);
+}
+
+void BasicBlock::addInstructionAtStart(InstBase *inst)
+{
+    _instructions.push_front(inst);
+}
+
+void BasicBlock::addInstructionAt(
+    InstBase *inst, std::list<InstBase *>::iterator it
+)
+{
+    _instructions.insert(it, inst);
+}
+
+void BasicBlock::addInstructionBefore(InstBase *inst, InstBase *before)
+{
+    auto it = std::find(_instructions.begin(), _instructions.end(), before);
```

Comment:
that's inefficient — maybe that's why LLVM and Swift use intrusive lists instead of std::list
(llvm::ilist)